drv/gimlet-hf-server/src/main.rs

```diff
@@ -180,6 +290,226 @@ impl ServerImpl {
             HfMuxState::HostCPU => Err(HfError::NotMuxedToSP),
         }
     }
+
+    fn page_program_raw(&self, addr: u32, data: &[u8]) -> Result<(), HfError> {
+        self.set_and_check_write_enable()?;
+        self.qspi.page_program(addr, data);
+        self.poll_for_write_complete(None);
+        Ok(())
+    }
+
+    fn set_dev(&mut self, state: HfDevSelect) -> Result<(), HfError> {
+        // Return early if the dev select pin is missing
+        let dev_select_pin = self.dev_select_pin.ok_or(HfError::NoDevSelect)?;
+
+        self.check_muxed_to_sp()?;
+
+        let sys = sys_api::Sys::from(SYS.get_task_id());
+        match state {
+            HfDevSelect::Flash0 => sys.gpio_reset(dev_select_pin),
+            HfDevSelect::Flash1 => sys.gpio_set(dev_select_pin),
+        }
+
+        self.dev_state = state;
+        Ok(())
+    }
+
+    fn get_raw_persistent_data(
+        &mut self,
+    ) -> Result<RawPersistentData, HfError> {
+        self.check_muxed_to_sp()?;
+        let best = if self.dev_select_pin.is_some() {
+            let prev_slot = self.dev_state;
+
+            // At this point, all of the ways that `set_dev` could fail have
+            // been checked: we already called both `check_muxed_to_sp` and that
+            // `dev_select_pin` is `Some(...)`, so we can unwrap returns from
+            // `self.set_dev(...)`.
+
+            // Look at the inactive slot first
+            self.set_dev(!prev_slot).unwrap();
+            let (a, _) = self.persistent_data_scan();
+
+            // Then switch back to our current slot, so that the resulting state
+            // is unchanged.
+            self.set_dev(prev_slot).unwrap();
+            let (b, _) = self.persistent_data_scan();
+
+            match (a, b) {
+                (Some(a), Some(b)) => {
+                    if a.epoch > b.epoch {
```

Comment:
Renamed `epoch`.  These should be equal unless something went wrong, because (a) the task ensure that they're equal on startup and (b) always writes them in pairs.  They could be unequal in practice if the second write of the pair fails and we haven't yet restarted the task.